File: cspro-dev/cspro/CSEntryDroid/app/src/main/java/gov/census/cspro/smartsync/p2p/BluetoothObexTransport.java
package gov.census.cspro.smartsync.p2p;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import android.bluetooth.BluetoothSocket;
import gov.census.cspro.util.CancelableTaskRunner;

public class BluetoothObexTransport {

	private BluetoothSocket m_socket;
	private final CancelableTaskRunner m_runner = new CancelableTaskRunner();
	private final ExecutorService m_threadPool = Executors.newSingleThreadExecutor();
	
	public BluetoothObexTransport(BluetoothSocket socket) {
		m_socket = socket;
	}
	
	public void close() {
		try {
			m_socket.close();
		} catch (IOException e) {
		}
	}
	
	public void write(byte[] data) throws IOException {
		m_socket.getOutputStream().write(data);
	}
	
	public int read(byte[] buffer, int byteCount, int timeoutMs) throws IOException {
		try {
			return m_runner.run(m_threadPool, 
					new ReadTask(m_socket, buffer, byteCount),
					null, timeoutMs, TimeUnit.MILLISECONDS);
		} catch (ExecutionException e) {
			throw new IOException(e);
		}
	}
	
	private static class ReadTask implements CancelableTaskRunner.CancelableTask<Integer> {

		private final BluetoothSocket m_socket;
		private final byte[] m_buffer;
		private final int m_byteCount;
		
		ReadTask(BluetoothSocket socket, byte[] buffer, int byteCount) {
			m_socket = socket;
			m_buffer = buffer;
			m_byteCount = byteCount;
		}
		
		@Override
		public Integer run() throws IOException {
			return m_socket.getInputStream().read(m_buffer, 0, m_byteCount);
		}

		@Override
		public void abort() {
			// Closing the socket will cause the synchronous read to
			// throw an exception and terminate.
			try {
				m_socket.close();
			} catch (IOException e) {
			}
		}
	}
}

File: src/wasmJsMain/kotlin/gov/census/cspro/smartsync/p2p/BluetoothObexTransport.kt
package gov.census.cspro.smartsync.p2p

/**
 * WASM/Web implementation of BluetoothObexTransport
 * 
 * Note: Web Bluetooth has significant limitations compared to Android Bluetooth.
 * OBEX protocol is not directly supported in browsers, and Web Bluetooth API
 * requires user gesture and HTTPS, with limited device/service support.
 * 
 * This is a stub implementation that logs appropriate messages when Bluetooth
 * operations are attempted. Full Bluetooth OBEX support would require:
 * - Web Bluetooth API (Chrome only, limited support)
 * - Custom GATT service implementation
 * - User gesture for device pairing
 * 
 * Mirrors: CSEntryDroid/app/src/main/java/gov/census/cspro/smartsync/p2p/BluetoothObexTransport.java
 */
actual class BluetoothObexTransport actual constructor(socket: BluetoothSocket) {
    private val socket = socket
    private var closed = false
    
    // Buffer for received data (stub - not actually used)
    private val receiveBuffer = mutableListOf<Byte>()
    
    actual fun close() {
        if (!closed) {
            closed = true
            socket.close()
            receiveBuffer.clear()
            println("[BluetoothObexTransport] Closed")
        }
    }
    
    actual suspend fun write(data: ByteArray) {
        if (closed) {
            throw Exception("Transport is closed")
        }
        
        println("[BluetoothObexTransport] Write requested for ${data.size} bytes")
        println("[BluetoothObexTransport] WARNING: Bluetooth OBEX not supported in browser environment")
        
        // In a real implementation, this would use Web Bluetooth GATT
        // For now, throw an exception indicating lack of support
        throw Exception("Bluetooth OBEX transport is not supported in the browser. Use network sync instead.")
    }
    
    actual suspend fun read(buffer: ByteArray, byteCount: Int, timeoutMs: Int): Int {
        if (closed) {
            throw Exception("Transport is closed")
        }
        
        println("[BluetoothObexTransport] Read requested for up to $byteCount bytes with ${timeoutMs}ms timeout")
        println("[BluetoothObexTransport] WARNING: Bluetooth OBEX not supported in browser environment")
        
        // In a real implementation, this would use Web Bluetooth GATT
        // For now, throw an exception indicating lack of support
        throw Exception("Bluetooth OBEX transport is not supported in the browser. Use network sync instead.")
    }
    
    companion object {
        /**
         * Check if Bluetooth is available in the current browser environment
         */
        fun isSupported(): Boolean {
            // Web Bluetooth is only available in secure contexts (HTTPS) and 
            // only in certain browsers (primarily Chrome)
            println("[BluetoothObexTransport] Checking Bluetooth support...")
            println("[BluetoothObexTransport] Web Bluetooth has limited browser support")
            return false // Stub: report as not supported
        }
    }
}

/**
 * WASM/Web implementation of BluetoothSocket
 * 
 * This is a stub implementation. Web Bluetooth sockets would need to wrap
 * a BluetoothDevice and BluetoothRemoteGATTServer from the Web Bluetooth API.
 */
actual class BluetoothSocket(
    private val deviceId: String = "unknown",
    private val deviceName: String = "Unknown Device"
) {
    private var closed = false
    private var connected = false
    
    actual fun isConnected(): Boolean {
        return !closed && connected
    }
    
    actual fun close() {
        if (!closed) {
            closed = true
            connected = false
            println("[BluetoothSocket] Closed - device: $deviceId")
        }
    }
    
    actual fun getRemoteAddress(): String {
        return deviceId
    }
    
    /**
     * Get the device name
     */
    fun getDeviceName(): String {
        return deviceName
    }
    
    companion object {
        /**
         * Create a stub socket for testing or when Bluetooth is not available
         */
        fun createStub(deviceId: String = "stub-device", deviceName: String = "Stub Device"): BluetoothSocket {
            println("[BluetoothSocket] Creating stub socket - Bluetooth not supported in browser")
            return BluetoothSocket(deviceId, deviceName)
        }
    }
}
